lib/imgbundler/imgbundler.go

```diff
@@ -0,0 +1,163 @@
+package imgbundler
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/multierr"
+	"oss.terrastruct.com/xdefer"
+
+	"oss.terrastruct.com/d2/lib/xmain"
+)
+
+const maxImageSize int64 = 1 << 25 // 33_554_432
+
+var imageRe = regexp.MustCompile(`<image href="([^"]+)"`)
+
+type resp struct {
+	srctxt string
+	data   string
+	err    error
+}
+
+func InlineLocal(ctx context.Context, ms *xmain.State, in []byte) ([]byte, error) {
+	return inline(ctx, ms, in, false)
+}
+
+func InlineRemote(ctx context.Context, ms *xmain.State, in []byte) ([]byte, error) {
+	return inline(ctx, ms, in, true)
+}
+
+func inline(ctx context.Context, ms *xmain.State, svg []byte, isRemote bool) (_ []byte, err error) {
+	defer xdefer.Errorf(&err, "failed to bundle images")
+	imgs := imageRe.FindAllSubmatch(svg, -1)
+
+	var filtered [][][]byte
+	for _, img := range imgs {
+		u, err := url.Parse(string(img[1]))
+		isRemoteImg := err == nil && strings.HasPrefix(u.Scheme, "http")
+		if isRemoteImg == isRemote {
+			filtered = append(filtered, img)
+		}
+	}
+
+	var wg sync.WaitGroup
+	respChan := make(chan resp)
+	// Limits the number of workers to 16.
+	sema := make(chan struct{}, 16)
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*5)
+	defer cancel()
+
+	wg.Add(len(filtered))
+	// Start workers as the sema allows.
+	go func() {
+		for _, img := range filtered {
+			sema <- struct{}{}
+			go func(src, href string) {
+				defer func() {
+					wg.Done()
+					<-sema
+				}()
+
+				var data string
+				var err error
+				if isRemote {
+					data, err = fetch(ctx, href)
+				} else {
+					data, err = read(href)
+				}
+				select {
+				case <-ctx.Done():
+				case respChan <- resp{
+					srctxt: src,
+					data:   data,
+					err:    err,
+				}:
+				}
+			}(string(img[0]), string(img[1]))
+		}
+	}()
+
+	go func() {
+		wg.Wait()
+		close(respChan)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			ms.Log.Debug.Printf("there")
+			return nil, fmt.Errorf("failed to wait for imgbundler workers: %w", ctx.Err())
+		case <-time.After(time.Second * 5):
+			ms.Log.Info.Printf("fetching images...")
+		case resp, ok := <-respChan:
+			if !ok {
+				return svg, err
+			}
+			if resp.err != nil {
+				err = multierr.Combine(err, resp.err)
```

Comment:
Remove `multierr` and log the errors. Otherwise `cmdlog.Logger.Nerrors` will always return zero.

Reply:
i changed the `return svg, nil` to `return svg, err` above.

Reply:
Oh nice catch but I'm saying we get rid of multierr to log all errors as they come in instead of waiting for all images to be fetched before we log errors. Otherwise there's no point in using `Nerrors`, you could just set a bool in compile() when the imgBundler returns an error and then return an xmain.ExitError with with status 1 if said bool is true at the end of the function.

Reply:
hmmm, then this whole thing just doesn't return error at all?

Reply:
nvm

Reply:
actually yeah, then there'd be no point in returning an error. which is odd. i think the current way is fine. Nerrors is just checking > 0 anyway, so it's all the same. yes it's the same as setting a bool, but it's cleaner, no extra var

Reply:
I don't see why that's odd? It means the CLI is more responsive.

Or actually, yea you ought to return a boolean here then and not even rely on `Nerrors`, I'm going to remove that. It's not cleaner, it's an unnecessary/nonlocal/magic API.

i.e return an ok bool from this function when the svg is totally good and no images errored and return false when at least one image failed. and then use that bool in compile to return the exit error.

Reply:
ok sure. but why return a bool? why not just return the error like before?

Reply:
I figure you don't need to, you're going to log that error, you're just going to check if it exists so a simple boolean suffices.

But now that I think about it, just return the `xmain.ExitError` here and then in `compile()` return the return error of `imgbundler.Inline<whatever>` after writing the output. That's perfect.